tests_e2e/tests/agent_update/rsm_update.py

```diff
@@ -0,0 +1,189 @@
+
+# Microsoft Azure Linux Agent
+#
+# Copyright 2018 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#
+# BVT for the agent update scenario
+#
+# The test verifies agent update for rsm workflow. This test covers three scenarios downgrade, upgrade and no update.
+# For each scenario, we intiaite the rsm request with target version and then verify agent updated to that target version.
+#
+import json
+from typing import List, Dict, Any
+
+import requests
+from azure.identity import DefaultAzureCredential
+from azure.mgmt.compute.models import VirtualMachine
+
+from tests_e2e.tests.lib.agent_test import AgentTest
+from tests_e2e.tests.lib.agent_test_context import AgentTestContext
+from tests_e2e.tests.lib.logging import log
+from tests_e2e.tests.lib.retry import retry_if_not_found
+from tests_e2e.tests.lib.ssh_client import SshClient
+from tests_e2e.tests.lib.virtual_machine import VmMachine
+
+
+class RsmUpdateBvt(AgentTest):
+
+    def __init__(self, context: AgentTestContext):
+        super().__init__(context)
+        self._ssh_client = SshClient(
+            ip_address=self._context.vm_ip_address,
+            username=self._context.username,
+            private_key_file=self._context.private_key_file)
+
+    def get_ignore_error_rules(self) -> List[Dict[str, Any]]:
+        ignore_rules = [
+            #
+            # This is expected as we validate the downgrade scenario
+            #
+            # WARNING ExtHandler ExtHandler Agent WALinuxAgent-9.9.9.9 is permanently blacklisted
+            #
+            {
+                'message': r"Agent WALinuxAgent-9.9.9.9 is permanently blacklisted"
+            }
+
+        ]
+        return ignore_rules
+
+    def run(self) -> None:
+        # Allow agent to send supported feature flag
+        self._verify_agent_reported_supported_feature_flag()
+
+        log.info("*******Verifying the Agent Downgrade scenario*******")
+        self._mock_rsm_update("1.3.0.0")
+        self._check_rsm_gs("1.3.0.0")
+        self._prepare_agent()
+
+        # Verify downgrade scenario
+        self._verify_guest_agent_update("1.3.0.0")
+
+        # Verify upgrade scenario
+        log.info("*******Verifying the Agent Upgrade scenario*******")
+        self._mock_rsm_update("1.3.1.0")
+        self._check_rsm_gs("1.3.1.0")
+        self._verify_guest_agent_update("1.3.1.0")
+
+        # verify no version update. There is bug in CRP and will enable once it's fixed
```

Comment:
what is the current bug?

Reply:
CRP not populating GA family in GS if requested version is same as last successful attempted updated version.

Before I was on vacation, Kashif mentioned change was merged, waiting for next CRP deployment